packages/backend-common/src/reading/GoogleGcsUrlReader.ts

```diff
@@ -110,8 +110,34 @@ export class GoogleGcsUrlReader implements UrlReader {
     throw new Error('GcsUrlReader does not implement readTree');
   }
 
-  async search(): Promise<SearchResponse> {
-    throw new Error('GcsUrlReader does not implement search');
+  async search(url: string): Promise<SearchResponse> {
+    const { bucket, key: pattern } = parseURL(url);
+
+    if (!pattern.endsWith('*') || pattern.indexOf('*') !== pattern.length - 1) {
+      throw new Error('GcsUrlReader only supports prefix-based searches');
+    }
+
+    const [files] = await this.storage.bucket(bucket).getFiles({
+      autoPaginate: true,
+      prefix: pattern.split('*').join(''),
+    });
+
+    return {
+      files: files.map(file => {
+        const fullUrl = ['https:/', GOOGLE_GCS_HOST, bucket, file.name].join(
+          '/',
+        );
+        return {
+          url: fullUrl,
+          content: async () => {
+            const readResponse = await this.readUrl(fullUrl);
+            return readResponse.buffer();
+          },
+        };
+      }),
+      // TODO etag is not implemented yet.
+      etag: 'NOT/IMPLEMENTED',
```

Comment:
I guess there's no implications of having this hard coded. Isn't there some field in the metadata response from `.getFiles`?

Reply:
There's plenty of metadata returned on a file-by-file basis, limited search-level metadata, unfortunately.  Bigger question from me, though: what does the `etag` _represent_ on a search?  And is that concept applicable across all `UrlReader` implementations?

Like if I were to artificially generate a value for this, when should it change?  When the list of names changes?  When the list of names is the same, but modification date is different?

Reply:
Good question!
(additional context probably redundant for you but any other reader) We initially added etag to be able to send it to the upstream service in order to cut down on API calls and quota meaning ideally we should send the last used etag to the upstream and throw a NotModified error if the etag is the same. This means the caller need use the last item from cache or otherwise decide what to do.

If possible I still think it would be good to return the etag from the wildcard response and also sending it in the outgoing request to google. Any form of internal compute probably based on the files probably defeats that purpose.

Reply:
Makes sense.  Yeah, unfortunately [the response from the API](https://googleapis.dev/nodejs/storage/latest/global.html#GetFilesResponse) leaves quite a bit to be desired.  Nothing we can really use as an `etag` equivalent.

Would it be useful to make the semantics of the value the same for callers?  e.g. just send a randomly generated UUID that will never match a future call.  ...Or does it not really make a difference and this `NOT/IMPLEMENTED` is fine?

Reply:
Well that response is not very helpful at all! :( I checked the node types and it looked like there was a metadata object next to files part of the response object. Nothing interesting in that one?

If the answer is no to the above then I think the `NOT/IMPLEMENTED` is sufficient 👍

Reply:
Yeah, unfortunately that `metadata` object is always `undefined`.

Reply:
Anything else @jhaals?  Otherwise, may I get a ✅ so I can :shipit: ?  😬 😬 😬

Reply:
Sorry I missed your message yesterday LGTM!